feat(std/fs): add support to parse home directory from /etc/passwd

jobansd opened the pull request:
Implemented the TODO in [`get_app_data_dir.zig`](lib/std/fs/get_app_data_dir.zig) that suggests parsing the home directory from `/etc/passwd`.

### Changes

- Added `getHomeDirFromPasswd()` function, which implements the parsing logic.
- ~~Changed the return type of `getAppDataDir()` from `GetAppDataDirError![]u8` to `![]u8` to encompass errors from `fs.openFileAbsolute()` and `reader.read()`~~
- `GetAppDataDirError` retained in return type, updated the `getHomeDirFromPasswd()` function call to catch and return `error.AppDataDirUnavailable` instead (suggested by @squeek502).
- Added declaration for `getuid()` to [`c.zig`](lib/std/c.zig).
- Added wrapper function for `getuid()` to [`posix.zig`](lib/std/posix.zig).
- Changed `std.os.linux.getuid()` to `std.posix.getuid()` in [`get_app_data_dir.zig`](lib/std/fs/get_app_data_dir.zig).

jobansd committed:
feat(std/fs): add support to parse home directory from /etc/passwd

Implemented the TODO in `get_app_data_dir.zig` that suggests parsing the home directory from `/etc/passwd`.

squeek502 commented:
IMO the `GetAppDataDirError` error set should be retained, and the `getHomeDirFromPasswd` calls should be changed to

```zig
getHomeDirFromPasswd() catch return error.AppDataDirUnavailable
```

so something like

```zig
const home_dir = posix.getenv("HOME") orelse (getHomeDirFromPasswd() catch return error.AppDataDirUnavailable);
```
or
```zig
const home_dir = posix.getenv("HOME") orelse blk: {
    break :blk getHomeDirFromPasswd() catch return error.AppDataDirUnavailable;
};
```

jobansd committed:
refactor(std/fs): retain GetAppDataDirError in return type

Restored `GetAppDataDirError` error set in return type.

jobansd commented:
@squeek502, thanks for the suggestion. I agree with you completely and I've pushed the change.

squeek502 commented on lib/std/fs/get_app_data_dir.zig:
```diff
@@ -57,6 +51,101 @@ pub fn getAppDataDir(allocator: mem.Allocator, appname: []const u8) GetAppDataDi
+/// Parses home directory from /etc/passwd.
+fn getHomeDirFromPasswd() ![]u8 {
+    const ReaderState = enum {
+        Start,
```
nitpick: these enum fields should be `snake_case`: https://ziglang.org/documentation/master/#Names

jobansd committed:
refactor(std/fs): move getuid() call out of while loop

Moved `std.os.linux.getuid()` call out of the while loop to reduce invocations and created variable `currentUid` to store the value.

jobansd committed:
refactor(std/fs): snake_case ReaderState enum fields

Changed fields in ReaderState enum to snake_case.

jobansd commented on lib/std/fs/get_app_data_dir.zig:
```diff
@@ -57,6 +51,102 @@ pub fn getAppDataDir(allocator: mem.Allocator, appname: []const u8) GetAppDataDi
+
+    var buf: [std.heap.page_size_min]u8 = undefined;
+    var state = ReaderState.start;
+    const currentUid = std.os.linux.getuid();
```
Good question, I could not confidently answer that. All I know is `getuid()` is a part of [`libc`](https://man7.org/linux/man-pages/man2/getuid.2.html). However, the function definition in the Zig [source](https://github.com/ziglang/zig/blob/03123916e55a9c0c8b66f748a62b7a4a64203535/lib/std/os/linux.zig#L1601):
```zig
pub fn getuid() uid_t {
    if (@hasField(SYS, "getuid32")) {
        return @as(uid_t, @intCast(syscall0(.getuid32)));
    } else {
        return @as(uid_t, @intCast(syscall0(.getuid)));
    }
}
```
makes a system call regardless of if `libc` is linked or not. `getHomeDirFromPasswd()` is called on `macos`, `linux`, `freebsd`, `netbsd`, `dragonfly`, `openbsd`, `solaris`, `illumos`, `serenity` and from cursory search I did they all have their own `libc` implementations and man pages for `getuid()`. With that said my assumption was "If it had `libc` and you compiled with `libc` it should be good."

Truthfully, I'm a novice and would love feedback on if this train of thought made sense or if there were betters questions I could've asked myself.

squeek502 replied:
(see my last comment if you missed it)

Could also add a wrapper function to `std.posix`, something like:

```
pub fn getuid() uid_t {
    return system.getuid();
}
```

then change the call to `std.posix.getuid()`

jobansd replied:
> Might be worth waiting for a core Zig team member to confirm, though.

With that said, I wanted to suggest a deeper change that I'd be down to implement. It's useful that we have `getAppDataDir()`, but how'd y'all feel about about adding some of the other XDG directories ie. `getAppConfigDir()`, `getAppCacheDir()`, `getAppStateDir()`, etc. translated across operating systems. More specifically and importantly a `getUserHomeDir()` since most of these directories live there and `std.fs.path.join()` can be used from there. Would it live better under `os` or `process`.

PS. I can move this to Zulip if you'd like.

squeek502 replied:
My 2 cents is that the [known-folders](https://github.com/ziglibs/known-folders) library fills that use case. `getAppDataDir` only exists because the Zig compiler needs it.

jobansd committed:
refactor(std/fs): enhance portability of getHomeDirFromPasswd()

Replaced `std.os.linux.getuid()` with `std.posix.getuid()` and added supporting changes to enhance portability of `getHomeDirFromPasswd()`.

alexrp commented:
This code has been deleted from the standard library.
